File: sdk/go/hello.go
package graft

import (
	"encoding/binary"
	"fmt"
)

// HelloResp represents the parsed HELLO_RESP payload.
type HelloResp struct {
	FirmwareName    string
	FirmwareVersion [3]byte // major, minor, patch
	McuID           []byte
	OTACapable      bool
	Pins            []PinCapability
	I2CBuses        byte
	SPIBuses        byte
	UARTCount       byte
	MaxPayload      uint16
	Modules         []ModuleDescriptor
	Datastreams     []DatastreamDescriptor
}

// PinCapability describes a single pin's capabilities.
type PinCapability struct {
	Pin          byte
	Capabilities byte
}

// ModuleDescriptor describes a loaded firmware module.
type ModuleDescriptor struct {
	ModuleID     byte
	Name         string
	VersionMajor byte
	VersionMinor byte
	Pins         []byte
}

// DatastreamDescriptor describes a datastream.
type DatastreamDescriptor struct {
	Index    byte
	Name     string
	Type     byte
	Unit     string
	Writable bool
	PinRef   byte
	Retain   bool
}

func readFixedString(data []byte, offset, length int) (string, int) {
	if offset+length > len(data) {
		return "", offset + length
	}
	raw := data[offset : offset+length]
	end := -1
	for i, b := range raw {
		if b == 0 {
			end = i
			break
		}
	}
	if end == -1 {
		end = length
	}
	return string(raw[:end]), offset + length
}

// ParseHelloResp parses a HELLO_RESP binary payload.
func ParseHelloResp(payload []byte) (*HelloResp, error) {
	if len(payload) < 30 {
		return nil, fmt.Errorf("hello payload too short: %d bytes", len(payload))
	}

	pos := 0
	var name string
	name, pos = readFixedString(payload, pos, 16)

	vMajor := payload[pos]
	pos++
	vMinor := payload[pos]
	pos++
	vPatch := payload[pos]
	pos++

	mcuID := make([]byte, 8)
	copy(mcuID, payload[pos:pos+8])
	pos += 8

	otaCapable := payload[pos] == 0x01
	pos++

	pinCount := int(payload[pos])
	pos++
	pins := make([]PinCapability, pinCount)
	for i := 0; i < pinCount; i++ {
		pins[i] = PinCapability{Pin: byte(i), Capabilities: payload[pos]}
		pos++
	}

	i2cBuses := payload[pos]
	pos++
	spiBuses := payload[pos]
	pos++
	uartCount := payload[pos]
	pos++

	maxPayload := binary.LittleEndian.Uint16(payload[pos:])
	pos += 2

	moduleCount := int(payload[pos])
	pos++
	modules := make([]ModuleDescriptor, moduleCount)
	for i := 0; i < moduleCount; i++ {
		modID := payload[pos]
		pos++
		var modName string
		modName, pos = readFixedString(payload, pos, 8)
		verMajor := payload[pos]
		pos++
		verMinor := payload[pos]
		pos++
		modPinCount := int(payload[pos])
		pos++
		modPins := make([]byte, modPinCount)
		for p := 0; p < modPinCount; p++ {
			modPins[p] = payload[pos]
			pos++
		}
		modules[i] = ModuleDescriptor{
			ModuleID: modID, Name: modName,
			VersionMajor: verMajor, VersionMinor: verMinor, Pins: modPins,
		}
	}

	dsCount := int(payload[pos])
	pos++
	datastreams := make([]DatastreamDescriptor, dsCount)
	for i := 0; i < dsCount; i++ {
		var dsName string
		dsName, pos = readFixedString(payload, pos, 16)
		dsType := payload[pos]
		pos++
		var dsUnit string
		dsUnit, pos = readFixedString(payload, pos, 8)
		writable := payload[pos] == 0x01
		pos++
		pinRef := payload[pos]
		pos++
		retain := payload[pos] == 0x01
		pos++
		datastreams[i] = DatastreamDescriptor{
			Index: byte(i), Name: dsName, Type: dsType, Unit: dsUnit,
			Writable: writable, PinRef: pinRef, Retain: retain,
		}
	}

	return &HelloResp{
		FirmwareName:    name,
		FirmwareVersion: [3]byte{vMajor, vMinor, vPatch},
		McuID:           mcuID,
		OTACapable:      otaCapable,
		Pins:            pins,
		I2CBuses:        i2cBuses,
		SPIBuses:        spiBuses,
		UARTCount:       uartCount,
		MaxPayload:      maxPayload,
		Modules:         modules,
		Datastreams:     datastreams,
	}, nil
}

File: sdk/js/src/hello.ts
/**
 * GRAFT HELLO_RESP Parser
 *
 * Decodes the binary HELLO_RESP payload into a typed capability model.
 * This is the host SDK's source of truth for what the device can do.
 */

import type { HelloResp, PinCapability, ModuleDescriptor, DatastreamDescriptor } from './core/types.js'

/**
 * Read a null-padded fixed-length string from a DataView.
 */
function readFixedString(view: DataView, offset: number, len: number): [string, number] {
  const bytes = new Uint8Array(view.buffer, view.byteOffset + offset, len)
  let end = bytes.indexOf(0)
  if (end === -1) end = len
  const str = new TextDecoder().decode(bytes.subarray(0, end))
  return [str, offset + len]
}

/**
 * Parse HELLO_RESP binary payload into a typed HelloResp object.
 */
export function parseHelloResp(payload: Uint8Array): HelloResp {
  const view = new DataView(payload.buffer, payload.byteOffset, payload.byteLength)
  let pos = 0

  // firmware_name: 16 bytes null-padded
  let firmwareName: string
  ;[firmwareName, pos] = readFixedString(view, pos, 16)

  // firmware_version: 3 bytes
  const vMajor = view.getUint8(pos++)
  const vMinor = view.getUint8(pos++)
  const vPatch = view.getUint8(pos++)

  // mcu_id: 8 bytes
  const mcuId = new Uint8Array(payload.buffer, payload.byteOffset + pos, 8)
  pos += 8

  // ota_capable: 1 byte
  const otaCapable = view.getUint8(pos++) === 0x01

  // pin_count: 1 byte
  const pinCount = view.getUint8(pos++)

  // pins[]: per-pin capability bitmask
  const pins: PinCapability[] = []
  for (let i = 0; i < pinCount; i++) {
    pins.push({ pin: i, capabilities: view.getUint8(pos++) })
  }

  // i2c_buses: 1 byte
  const i2cBuses = view.getUint8(pos++)

  // spi_buses: 1 byte
  const spiBuses = view.getUint8(pos++)

  // uart_count: 1 byte
  const uartCount = view.getUint8(pos++)

  // max_payload: uint16 LE
  const maxPayload = view.getUint16(pos, true)
  pos += 2

  // module_count: 1 byte
  const moduleCount = view.getUint8(pos++)

  // modules[]
  const modules: ModuleDescriptor[] = []
  for (let i = 0; i < moduleCount; i++) {
    const moduleId = view.getUint8(pos++)

    let moduleName: string
    ;[moduleName, pos] = readFixedString(view, pos, 8)

    const versionMajor = view.getUint8(pos++)
    const versionMinor = view.getUint8(pos++)

    const modPinCount = view.getUint8(pos++)
    const modPins: number[] = []
    for (let p = 0; p < modPinCount; p++) {
      modPins.push(view.getUint8(pos++))
    }

    modules.push({ moduleId, name: moduleName, versionMajor, versionMinor, pins: modPins })
  }

  // datastream_count: 1 byte
  const dsCount = view.getUint8(pos++)

  // datastreams[]
  const datastreams: DatastreamDescriptor[] = []
  for (let i = 0; i < dsCount; i++) {
    let dsName: string
    ;[dsName, pos] = readFixedString(view, pos, 16)

    const dsType = view.getUint8(pos++)

    let dsUnit: string
    ;[dsUnit, pos] = readFixedString(view, pos, 8)

    const writable = view.getUint8(pos++) === 0x01
    const pinRef = view.getUint8(pos++)
    const retain = view.getUint8(pos++) === 0x01

    datastreams.push({
      index: i,
      name: dsName,
      type: dsType,
      unit: dsUnit,
      writable,
      pinRef,
      retain,
    })
  }

  return {
    firmwareName,
    firmwareVersion: [vMajor, vMinor, vPatch],
    mcuId: new Uint8Array(mcuId),
    otaCapable,
    pins,
    i2cBuses,
    spiBuses,
    uartCount,
    maxPayload,
    modules,
    datastreams,
  }
}
